mreg_cli/api/models.py

```diff
@@ -255,30 +262,58 @@ def is_delegated(self) -> bool:
         return True
 
 
-class Role(FrozenModelWithTimestamps, APIMixin["Role"]):
-    """Model for a role.
+class HostPolicyBase(FrozenModelWithTimestamps):
+    """Base model for Host Policy objects.
+
+    Note:
+    ----
+    Host policy models in MREG have a different `created_at` field than
+    other models. It is called `create_date` and is a date - not a datetime.
+
+    This model has a custom validator to validate and convert the `create_date`
+    field to a datetime object with the expected `created_at` name.
 
-    Note that HostPolicy throws out `create_date` as the date only, not as a
-    proper datetime object, and not with the expected name `created_at`.
     """
 
-    id: int  # noqa: A003
     created_at: datetime = Field(..., validation_alias=AliasChoices("create_date", "created_at"))
-    hosts: NameList
-    atoms: NameList
-    description: str
     name: str
-    labels: list[int]
+    description: str
 
     @field_validator("created_at", mode="before")
     @classmethod
-    def validate_created_at(cls, value: str) -> datetime:
-        """Validate and convert the created_at field to datetime.
+    def validate_created_at(cls, value: Any) -> datetime:
+        """Convert a datetime string to a datetime object.
 
-        :param value: Input date string from the JSON.
-        :returns: Converted datetime object.
+        :param value: The input value - should be a datetime string.
+        :returns: The input value converted to a datetime object.
         """
-        return datetime.fromisoformat(value)
+        # Fast path for str (most likely input type)
+        if isinstance(value, str):
+            return datetime.fromisoformat(value)
+        elif isinstance(value, datetime):
+            return value
+        elif isinstance(value, date):
+            return datetime.combine(value, datetime.min.time())
+        return value  # let pydantic throw the ValidationError
+
+    @property
+    def created_at_with_tz(self) -> datetime:
+        """The constructed `created_at` field with timezone info from `updated_at`."""
+        return self.created_at.replace(tzinfo=self.updated_at.tzinfo)
```

Comment:
Need a better way to handle this. We can't use an after validator on the model itself due to the fields being frozen. We could assign the `created_at` field to a hidden field, and then add a `@computed_field` called `create_date` which returns a tz aware create time.

Reply:
At some point this feels like something we should fix on the server side, in the database.

Reply:
I was hoping this would work, but you cannot override normal fields with computed fields:

```py
class HostPolicyBase(FrozenModelWithTimestamps):
    # ...
    # The constructed datetime object without timezone info.
    created_at_tz_naive: datetime = Field(
        ...,
        validation_alias=AliasChoices("create_date", "created_at", "created_at_tz_naive"),
        exclude=True,
    )
    # ...

    @computed_field
    def created_at(self) -> datetime:
        """The constructed `created_at` datetime object with timezone info injected from `updated_at`."""
        return self.created_at_tz_naive.replace(tzinfo=self.updated_at.tzinfo)

    @computed_field
    def create_date(self) -> date:
        """Original `create_date` field."""
        return self.created_at_tz_naive.date()
```

```
  File "/Users/pederhan/repos/mreg-cli/venv/lib/python3.11/site-packages/pydantic/_internal/_model_construction.py", line 474, in set_model_fields
    fields, class_vars = collect_model_fields(cls, bases, config_wrapper, types_namespace, typevars_map=typevars_map)
                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/pederhan/repos/mreg-cli/venv/lib/python3.11/site-packages/pydantic/_internal/_fields.py", line 242, in collect_model_fields
    raise ValueError("you can't override a field with a computed field")
ValueError: you can't override a field with a computed field
````

Reply:
I guess we'll just do the "dumb" thing and leave it in a dangerous state and add a very stern comment about using it for comparisons.

Reply:
>  You can't override a field from a parent class with a computed_field in the child class. mypy complains about this behavior if allowed, and dataclasses doesn't allow this pattern either.

_From the docstring of `computed_field`._ 

So this strictness from Pydantic's side is just a result of type-checking limitations and wanting consistency with dataclasses 🤦

Reply:
Well, we could rename the field `_created_at` and then make the `@property` `created_at`? I mean... We could?

Reply:
Fields can't be prefixed with underscores; Pydantic throws an exception if you try to assign `Field(...)` to an attribute name with a leading underscore. 

But yeah, I was playing with the idea of doing something like that, the only "problem" is that we have to customize model serialization with `@model_serializer` if we want to include the property if/when we serialize the model (maybe not very relevant?), hence why I was trying to make it work with `@computed_field`.

Reply:
I don't really see the serialisation gain at the moment, as we're only delivering human-targeted output, and if that becomes an issue, we can deal with it then (famous last words). `create_date_not_to_be_used_stop_thinking_about_it_I_am_broken` would be an option. :)

Reply:
Overriding a field with a `@property` when subclassing does not work. It returns the property decorator instead of the value. 

Instead of hacking together something brittle, I have changed `HostPolicy` to inherit from `FrozenModel`, and then implemented the `created_at` and `updated_at` fields directly on the model, so we can be sure this special case works. Copy-pasting `output_timestamps` is NOT ideal, but we can refactor it to use a common function if we decide to use this approach.